src/build.rs

```diff
@@ -64,15 +66,62 @@ impl BuildSystem {
         }
     }
 
+    /// Creates a "staging area" (dist/.stage) for storing intermediate build results
+    async fn prepare_dist_staging(&self) -> Result<()> {
+        // Prepare staging area in which we will assemble the latest build
+        let dist_staging: &Path = self.cfg.dist_staging.as_path().into();
+
+        // Clean staging area, if applicable
+        remove_dir_all(dist_staging.into()).await.context("error cleaning staging dist dir")?;
+        fs::create_dir_all(dist_staging)
+            .await
+            .with_context(|| "error creating build environment directory: staging dist dir")?;
+
+        Ok(())
+    }
+
     async fn do_build(&mut self) -> Result<()> {
-        // TODO: delete the contents of the `dist/.current` dir (currently in flight elsewhere).
+        // Ensure the output dist directories are in place.
+        fs::create_dir_all(self.cfg.dist_final.as_path())
+            .await
+            .with_context(|| "error creating build environment directory: dist")?;
 
-        // Ensure the output dist directory is in place.
-        fs::create_dir_all(self.cfg.dist.as_path()).await?;
+        self.prepare_dist_staging().await.context("error preparing build environment")?;
 
         // Spawn the source HTML pipeline. This will spawn all other pipelines derived from
         // the source HTML, and will ultimately generate and write the final HTML.
         self.html_pipeline.clone().spawn().await?;
+
+        // Move distrbution from staging dist to final dist
+        self.finalize_dist().await.context("error applying built distribution")?;
+        Ok(())
+    }
+
+    /// Moves the contents of dist/.stage into dist, signifying the application
+    /// of a successful build. Also removes dist/.stage afterwards.
+    async fn finalize_dist(&self) -> Result<()> {
+        let dist_final = self.cfg.dist_final.clone();
+        let dist_staging = self.cfg.dist_staging.clone();
+        self.progress.clone().set_message("applying new distribution");
+
+        // Build succeeded, so delete everything in `dist`,
+        // move everything from `dist/.stage` to `dist`, and
+        // then delete `dist/.stage`.
+        let mut entries = fs::read_dir(&dist_staging).await.context("error reading final dist dir")?;
```

Comment:
The algorithm below looks good, though I think we are missing the routine to clear out the contents of the dist dir (while preserving .stage obvs). I think that was there a few commits back.